packages/thrift-client/src/main/connections/HttpConnection.ts

```diff
@@ -217,56 +214,47 @@ export class HttpConnection extends Core.ThriftConnection<RequestOptions> {
                 : this.url
 
         // Merge user options with required options
-        const requestOptions: RequestOptions & UrlOptions = Core.overlayObjects(
-            this.requestOptions,
-            options,
+        const requestOptions: RequestOptions = mergeOptions(
+            mergeOptions(this.requestOptions, options),
             {
                 method: 'POST',
                 body: dataToWrite,
-                encoding: null, // Needs to be explicitly set to null to get Buffer in response body
-                url: requestUrl,
                 headers: {
-                    'Content-Length': dataToWrite.length,
                     'Content-Type': 'application/octet-stream',
                 },
             },
         )
 
         return new Promise((resolve, reject) => {
-            this.requestImpl(
-                requestOptions,
-                (err: any, response: RequestResponse, body: Buffer) => {
+            this.requestImpl(requestUrl, requestOptions)
+                .then(response => {
                     if (
                         shouldRetry(response, retry, this.withEndpointPerMethod)
                     ) {
                         resolve(
                             this.write(dataToWrite, methodName, options, true),
                         )
+                    } else if (isErrorResponse(response)) {
+                        reject(response)
                     } else {
-                        if (hasError(err)) {
-                            reject(err)
-                        } else if (isErrorResponse(response)) {
-                            reject(response)
-                        } else {
+                        response.arrayBuffer().then(arrayBuffer =>
                             resolve({
-                                statusCode: response.statusCode,
+                                statusCode: response.status,
                                 headers: response.headers,
-                                body,
-                            })
-                        }
+                                body: Buffer.from(arrayBuffer),
+                            }),
+                        )
                     }
-                },
-            )
+                })
+                .catch(e => reject(e))
```

Comment:
When I was testing, not doing this caused the exception to bubble out in a strange way and make the tests fail. I'll admit it felt a bit strange to me too (I initially did not have it), so I'll double check again that it's really necessary and see if there's a better explanation I can give you.

Reply:
I removed again and re-tested, things failed in some tests that were doing a `.then(successCallback, rejectCallback)` with both resolve and reject callbacks. I was reading [this article](https://dmitripavlutin.com/javascript-promises-then-vs-then-catch/), so maybe it's possible that the tests should change to use `.catch` instead of the `.then` with both callbacks? I guess we could have a breaking change here where someone expects their rejection callback to have this behavior and they'd have to update it to use `.catch`. So, I guess I'm in favor of leaving this as is, even if it seems strange. What do you think?

Reply:
Just for clairty:
```
        return new Promise((resolve, reject) => {
            this.requestImpl(requestUrl, requestOptions)
                .then(response => {...})
                .catch(e => reject(e))
```
we are creating a new `Promise` here that wraps a call to `requestImpl` If we do not have a `catch` for `requestImpl` that means this turns from a promise reject to a thrown error.  When making a promise you HAVE to call `reject` for consistent behavior. `new Promise` does NOT call `reject` automatically if an error is `thrown`. 

Also we should see this error: `rejecting a promise which was not handled with .catch().`
because `new Promise` does NOT make a promise chain.  only `.then()/.catch()` make chains. That means rejects are not propagated back "up the chain" because there is no chain in a `new Promise`

Reply:
Ohhh that makes sense. That's for the explanation. Then I guess that makes me wonder why we're wrapping `this.requestImpl()` in a new Promise. It looks like `fetch()` returns a promise so I'm wondering if this could be written as:
```typescript
     return this.requestImpl(requestUrl, requestOptions)
                .then(response => {...})
```

Reply:
ha yea, there is actually no reason to create a `new Promise` unless this project is using a custom `Promise` library like bluebird, but I don't think that is the case?

Reply:
Cool ok, thanks for helping me understand. I don't know if they're using a custom Promise library. I haven't looked. For the record I don't think it's important to spend time trying to improve this.